stubs/responses.pyi

```diff
@@ -0,0 +1,19 @@
+from typing import Any, Dict, Optional, TypeVar
+
+JsonDict = Dict[str, Any]
+
+DELETE: str
+GET: str
+POST: str
+PUT: str
+
+def add(
+    method: Optional[str] = None,
+    url: Optional[str] = None,
+    status: Optional[int] = None,
+    json: Optional[JsonDict] = None,
+): ...
+
+T = TypeVar("T")
+
+def activate(T) -> T: ...
```

Comment:
What does this def activate(T) function do?

Reply:
Context: we are using the `responses` library in our testing suite. This 3rd party library does not come with type annotations, so we have [a couple options](https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports) to deal with it:

1. We can simply tell `mypy` to ignore imports from `responses`.
2. We can write our own "stubs" to tell `mypy` what types to expect from the `responses` library.

(1) is easier but it means that `mypy` will let us pass anything to the `responses` library functions and won't complain.

(2) is a little fancier and means we have to write this `.pyi` file, but gives us nice type checking (e.g. `responses.add(asdf=False)` will complain that `asdf` is not a valid keyword to `responses.add` OR `responses.add(method=1)` will complain that `method` should be a `str`, not an `int`).

---

In the `responses` library, the `responses.activate` function is a [decorator](https://dbader.org/blog/python-decorators). Specifically, `responses.activate` takes a function as input and produce a function (with the same signature) as output.

We capture this knowledge here by saying, "for some arbitrary type `T`, the `activate` function should accept something of type `T` and produce something of type `T`".

Reply:
Decorator article is a very good read. Thanks for the link. It makes sense now.